runtime/src/utils/buffer/write.rs

```diff
@@ -2,104 +2,148 @@ use crate::{Blob, Error, RwLock};
-    position: u64,
+    /// This represents the logical position in the blob where `data[0]` would be written. The
+    /// buffer is maintained at the "tip" to support efficient size calculation and appends.
+    offset: u64,
     /// The maximum size of the buffer.
     capacity: usize,
 }
 
-impl<B: Blob> Inner<B> {
+impl Buffer {
+    /// Creates a new buffer with the provided `size` and `capacity`.
+    fn new(size: u64, capacity: usize) -> Self {
+        Self {
+            data: Vec::with_capacity(capacity),
+            offset: size,
+            capacity,
+        }
+    }
+
     /// Returns the current logical size of the blob including any buffered data.
     fn size(&self) -> u64 {
-        self.position + self.buffer.len() as u64
+        self.offset + self.data.len() as u64
     }
 
-    /// Appends bytes to the internal buffer, maintaining the "buffer at tip" invariant.
+    /// Returns true if the buffer is empty.
+    fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    /// Resizes the buffer to the provided `len`.
     ///
-    /// If the buffer capacity would be exceeded, it is flushed first. If the data
-    /// is larger than the buffer capacity, it is written directly to the blob.
+    /// If the new size is greater than the current size, the existing buffer is returned (to be flushed to
+    /// the underlying blob) and the buffer is reset to the empty state with an updated offset positioned at the end
+    /// of the logical blob.
     ///
-    /// Returns an error if the write to the underlying [Blob] fails (may be due to a `flush` of data not
-    /// related to the data being written).
-    async fn write<S: Into<StableBuf>>(&mut self, buf: S) -> Result<(), Error> {
-        // If the buffer capacity will be exceeded, flush the buffer first
-        let buf = buf.into();
-        let buf_len = buf.len();
-
-        // Flush buffer if adding this data would exceed capacity
-        if self.buffer.len() + buf_len > self.capacity {
-            self.flush().await?;
+    /// If the new size is less than the current size (but still greater than current offset), the buffer is truncated to the new size.
+    ///
+    /// If the new size is less than the current offset, the buffer is reset to the empty state with an updated offset positioned at the end
+    /// of the logical blob.
+    fn resize(&mut self, len: u64) -> Option<(Vec<u8>, u64)> {
+        // Handle case where the buffer is empty.
+        if self.is_empty() {
+            self.offset = len;
+            return None;
         }
 
-        // Write directly to blob if data is larger than buffer capacity
-        if buf_len > self.capacity {
-            self.blob.write_at(buf, self.position).await?;
-            self.position += buf_len as u64;
-            return Ok(());
+        // Handle case where there is some data in the buffer.
+        if len >= self.size() {
+            let previous = (
```

Comment:
why not just let r = self.take() ?

Reply:
Originally had that but then we update offset twice (one inside take and one here) 🤷

Reply:
Would likely get optimized away, in case you feel take() provides more clarity.

Reply:
Because it was over a function boundary, I was less sure.

In any case, I'll ponder.